File: Apps/Web/src/types/trade.ts
export interface TradeCreate {
  portfolio_id: string;
  asset_symbol: string;
  asset_name?: string;
  market_type?: string;
  exchange?: string;
  trade_type: "buy" | "sell";
  buy_price?: number;
  sell_price?: number;
  quantity: number;
  position_size?: number;
  fees_paid?: number;
  currency?: string;
  entry_at?: string;
  exit_at?: string;
  pnl_amount?: number;
  pnl_percentage?: number;
  notes?: string;
  strategy_tags?: string[];
}

export interface TradeUpdate {
  sell_price?: number;
  quantity?: number;
  fees_paid?: number;
  exit_at?: string;
  pnl_amount?: number;
  pnl_percentage?: number;
  notes?: string;
  strategy_tags?: string[];
}

export interface TradeRead {
  id: string;
  portfolio_id: string;
  asset_symbol: string;
  asset_name: string | null;
  market_type: string;
  exchange: string;
  trade_type: string;
  buy_price: string | null;
  sell_price: string | null;
  quantity: string;
  position_size: string | null;
  fees_paid: string;
  currency: string;
  entry_at: string | null;
  exit_at: string | null;
  pnl_amount: string | null;
  pnl_percentage: string | null;
  break_even_price: string | null;
  notes: string | null;
  strategy_tags: string[] | null;
  created_at: string;
}

export interface TradeListResponse {
  items: TradeRead[];
  count: number;
}

File: Apps/Api/schemas/trade.py
from __future__ import annotations
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field


class TradeCreate(BaseModel):
    portfolio_id: UUID
    asset_symbol: str = Field(..., min_length=1, max_length=30)
    asset_name: str | None = None
    market_type: str = "crypto"
    exchange: str = "Binance"
    trade_type: str = Field(..., pattern="^(buy|sell)$")
    buy_price: Decimal | None = Field(default=None, gt=0)
    sell_price: Decimal | None = Field(default=None, gt=0)
    quantity: Decimal = Field(..., gt=0)
    position_size: Decimal | None = Field(default=None, gt=0)
    fees_paid: Decimal = Field(default=Decimal("0"), ge=0)
    currency: str = "USDT"
    entry_at: datetime | None = None
    exit_at: datetime | None = None
    pnl_amount: Decimal | None = None
    pnl_percentage: Decimal | None = None
    break_even_price: Decimal | None = None
    notes: str | None = None
    strategy_tags: list[str] | None = None
    investment_amount: Decimal | None = Field(default=None, ge=0)
    category: str | None = Field(default=None, max_length=50)


class TradeUpdate(BaseModel):
    sell_price: Decimal | None = Field(default=None, gt=0)
    quantity: Decimal | None = Field(default=None, gt=0)
    fees_paid: Decimal | None = Field(default=None, ge=0)
    exit_at: datetime | None = None
    pnl_amount: Decimal | None = None
    pnl_percentage: Decimal | None = None
    notes: str | None = None
    strategy_tags: list[str] | None = None
    investment_amount: Decimal | None = Field(default=None, ge=0)
    category: str | None = Field(default=None, max_length=50)


class TradeRead(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    portfolio_id: UUID
    asset_symbol: str
    asset_name: str | None
    market_type: str
    exchange: str
    trade_type: str
    buy_price: Decimal | None
    sell_price: Decimal | None
    quantity: Decimal
    position_size: Decimal | None
    fees_paid: Decimal
    currency: str
    entry_at: datetime | None
    exit_at: datetime | None
    pnl_amount: Decimal | None
    pnl_percentage: Decimal | None
    break_even_price: Decimal | None
    notes: str | None
    strategy_tags: list[str] | None
    investment_amount: Decimal | None
    category: str | None
    created_at: datetime


class TradeListResponse(BaseModel):
    items: list[TradeRead]
    count: int
